### DISCUSSION TITLE
Can the implementation work with the voxellized data?

### DISCUSSION THREAD

**Discussion by EvgenyUgolkov** [octree-nn/ocnn-pytorch#27] [Q&A] [answered]
Dear Mr. Peng-Shuai Wang, good day!

I am working with the Computer-Tomography (CT-scan) images of rocks. These images are 3D voxel grids. For example, the image can be 1000x1000x1000 voxels of size, with each voxel to have a value of 0, 1, 2, or 3, etc. 

Large areas of the image can contain repetitive voxels. For example, the pores are only 0-s, and this way we have a big clusters of 0-s. Due to this structure, such images may also benefit a lot from the octree representation.

I was looking for the information if somebody implemented Conv3D layers for such octrees in Pytorch, and found this page. However, based on your paper, and based on the first analysis of the code, i can see that you are working with the 3D shapes, which assume coordinates and normals. My 3D voxel grids are different, but are they completely different?

Dear Mr. Peng-Shuai Wang, could you tell me, if the code you kindly provided here can work with such 3D voxel grids? If no, what is the shortest way for me to adapt it for the work with 3D voxel grids?

I try to implement the Super-Resolution algorithm, and Conv3D layers demand prohibitively big amount of GPU memory. For this reason, i am searching for the ways of optimization. For example, i start with 64x64x64 volumes, and up-sample them x8 times to the 512x512x512 voxels. Could you tell, is it possible to implement with octrees?

With respect, Evgeny

**Comment by harryseely** [comment]
Hey there! I have been using OCNN for almost 2 years now and I think you could probably adapt it to your project. All you would have to do is take the center coordinates of your voxel and convert this to a point cloud (XYZ) format.

**Reply by EvgenyUgolkov** [reply]
Dear Harry, hello!

Thank you for reaching me out! I ended up with Minkovsky Engine (https://github.com/NVIDIA/MinkowskiEngine). Maybe it could also useful for your research

**Reply by wang-ps** [reply, marked as answer]
Sorry, I missed the discussion. I agree that  @harryseely 's solution is OK, and MinkowskiEngine is also OK. The execution of convolution in ocnn and MinkowkiEngine is the same, they are exchangeable.

**Reply by EvgenyUgolkov** [reply]
@wang-ps wow, didn't expect to see you there. Thank you for the reply!

**Comment by mo12na34** [comment]
Hi @harryseely and @wang-ps, Can you tell me what is the final output of using OCNN? Like if you ask me about a task like image classification using CNN, then end goal is to predict the given image like Dog vs Cat, etc. So, what is the output using OCNN? It will be helpful if you can share a sample image of the predicted output.

**Comment by harryseely** [comment]
Hi @mo12na34 !

I am sure @wang-ps can provide a more detailed answer from the perspective of the developer. 

From my perspective as someone who has widely implemented OCNN in my work I would describe the output as follows:

- It depends on the architecture you are using. OCNN-SegNet will have a different output than OCNN-LeNet
- Typical neural networks are designed to output what are called "logits", which are tensors representing learned features about the data. These can be plugged into an added layer of the network for regression, classification, or whatever task you are performing. 
- I a have used OCNN-LeNet, OCNN-HRNet, and OCNN-ResNet for regressions tasks
- If you look at the LeNet model for example, the output is determined by the parameter "out_channels". The output shape is thus batch size (e.g., 32) x out_channels (e.g., 1024).
- If you want to see an example of how I tweaked OCNN for my task, you can have a look at my Github repo here:
https://github.com/harryseely/Biomass-DL/tree/main/pytorch_models
-I use the OCNN-HRNet architecture, and then pass the output logits to a smaller regressor module (simple MLP)

Happy to answer any other questions, big shoutout to PS-Wang for an awesome package!

-Harry

**Reply by mo12na34** [reply]
Thank you @harryseely

**Comment by wang-ps** [comment]
Thank @harryseely, for the detailed instructions. Yes, OCNN is a 3D CNN, and can be used for classification, segmentation, detection, and regression for 3D datasets.

**Reply by tech-ml** [reply]
Hi @wang-ps, I am a newbie and I am struct running this OCNN. Should I run all the commands across the different topics like, classification, segmentation, autoencoder, image2shape and finally completion? So, that I will get the desired output.

**Reply by wang-ps** [reply]
Yes, you can follow the documents to run the code. I would also suggest that you read the corresponding papers and the released PowerPoint slides to get more context on what's going on.

**Reply by tech-ml** [reply]
Okay, Thank you @wang-ps

**Reply by mo12na34** [reply]
Thank you @wang-ps
